File: scripts/build-data.py
"""Validate contributor JSON and deterministically compile the atlas. Python 3 only."""
import csv
import datetime as dt
import json
import math
import re
import sys
from pathlib import Path
from urllib.parse import urlsplit

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / 'data'
OUT = ROOT / 'public/data'


def read(name):
    def unique(pairs):
        result = {}
        for key, value in pairs:
            if key in result:
                raise ValueError(f'{name}: duplicate JSON key {key}')
            result[key] = value
        return result
    try:
        return json.loads(
            (DATA / name).read_text(encoding='utf-8'),
            object_pairs_hook=unique,
            parse_constant=lambda value: require(False, f'{name}: invalid number {value}'),
        )
    except json.JSONDecodeError as error:
        raise ValueError(f'{name}: {error}') from error


def require(condition, message):
    if not condition:
        raise ValueError(message)


def date_key(value):
    require(isinstance(value, str) and re.fullmatch(r'\d{4}(-\d{2}){0,2}', value), f'Invalid date: {value}')
    parts = [int(x) for x in value.split('-')]
    return dt.date(*(parts + [1] * (3 - len(parts))))


def record(value, required, context, optional=()):
    require(isinstance(value, dict), f'{context}: expected an object')
    missing = set(required) - value.keys()
    unknown = value.keys() - set(required) - set(optional)
    require(not missing, f'{context}: missing fields {sorted(missing)}')
    require(not unknown, f'{context}: unknown or generated fields {sorted(unknown)}')


def text(value, context):
    require(isinstance(value, str) and value.strip(), f'{context}: expected nonempty text')


def text_fields(value, names, context):
    for name in names:
        text(value[name], f'{context}.{name}')


def identifier(value, context):
    require(isinstance(value, str) and re.fullmatch(r'[a-z][a-z0-9-]*', value),
            f'{context}: expected a lowercase ID with letters, digits or hyphens')


def strings(value, context):
    require(isinstance(value, list) and value, f'{context}: expected a nonempty list')
    for item in value:
        text(item, context)
    require(len(value) == len(set(value)), f'{context}: duplicate entry')


def year(value, context):
    require(type(value) is int and 1000 <= value <= 9998, f'{context}: invalid year')


def records(value, required, context):
    require(isinstance(value, list) and value, f'{context}: expected a nonempty list')
    seen = set()
    for index, item in enumerate(value):
        label = f'{context}[{index}]'
        record(item, required, label)
        identifier(item['id'], f'{label}.id')
        require(item['id'] not in seen, f'{context}: duplicate ID {item["id"]}')
        seen.add(item['id'])


def compile_data():
    meta, owners, sources, contexts, scope = [
        read(f'{name}.json') for name in ['meta', 'owners', 'sources', 'contexts', 'scope']
    ]
    meta_text = ['title', 'version', 'scope', 'baseline', 'method', 'sovereignty',
                 'completeness', 'sourcePolicy', 'license', 'compiled', 'currentStatusAsOf']
    record(meta, [*meta_text, 'startYear', 'endYear'], 'meta.json')
    text_fields(meta, meta_text, 'meta.json')
    for field in ['compiled', 'currentStatusAsOf']:
        require(len(meta[field]) == 10, f'meta.{field}: use YYYY-MM-DD')
        date_key(meta[field])
    for field in ['startYear', 'endYear']:
        year(meta[field], f'meta.{field}')
    require(meta['startYear'] < meta['endYear'], 'meta: startYear must precede endYear')

    records(owners, ['id', 'label', 'description', 'color'], 'owners.json')
    records(sources, ['id', 'title', 'publisher', 'type', 'url', 'accessed'], 'sources.json')
    records(contexts, ['id', 'start', 'end', 'title', 'description', 'sources'], 'contexts.json')
    owner_ids = {owner['id'] for owner in owners}
    source_map = {source['id']: source for source in sources}
    for source in sources:
        text_fields(source, ['title', 'publisher', 'type', 'url', 'accessed'], source['id'])
        url = urlsplit(source['url'])
        require(url.scheme == 'https' and url.hostname and not url.username and not url.password,
                f'{source["id"]}: source needs an HTTPS URL without credentials')
        require(len(source['accessed']) == 10, f'{source["id"]}: accessed needs YYYY-MM-DD')
        date_key(source['accessed'])
    for owner in owners:
        text_fields(owner, ['label', 'description', 'color'], owner['id'])
        require(re.fullmatch(r'#[0-9a-fA-F]{6}', owner['color']),
                f'{owner["id"]}: invalid power colour')

    def citations(refs, context):
        strings(refs, f'{context}.sources')
        require(all(ref in source_map for ref in refs), f'{context}: unknown source in {refs}')

    for context in contexts:
        label = context['id']
        require(label not in {'all', 'custom'}, f'{label}: reserved date-preset ID')
        text_fields(context, ['title', 'description'], label)
        citations(context['sources'], label)
        year(context['start'], f'{label}.start')
        year(context['end'], f'{label}.end')
        require(meta['startYear'] <= context['start'] < context['end'] <= meta['endYear'],
                f'{label}: era must be within the atlas range')
    record(scope, ['places', 'tracks'], 'scope.json')
    strings(scope['tracks'], 'scope.tracks')
    strings(scope['places'], 'scope.places')
    for name in scope['tracks']:
        identifier(name, 'scope.tracks')
    files = {path.stem for path in (DATA / 'islands').glob('*.json')}
    require(files == set(scope['tracks']), 'Island files and scope.json tracks do not match')

    islands, all_event_ids = [], set()
    island_keys = ['id', 'name', 'place', 'region', 'coordinates', 'peoples', 'summary',
                   'notes', 'sources', 'initialController', 'initialSovereign', 'events']
    event_keys = ['id', 'date', 'precision', 'kind', 'title', 'detail', 'controller', 'sovereign', 'sources']
    optional = ['uncertainty', 'qualification', 'claimant']
    kinds = {'claim', 'context', 'resistance', 'status', 'capture', 'restoration',
             'settlement', 'treaty', 'independence', 'withdrawal', 'context-change', 'resistance-change'}
    for name in scope['tracks']:
        island = read(f'islands/{name}.json')
        record(island, island_keys, name)
        require(island['id'] == name, f'{name}: filename and ID differ')
        text_fields(island, ['name', 'place', 'region', 'peoples', 'summary', 'notes',
                             'initialController', 'initialSovereign'], name)
        require(island['place'] in scope['places'], f'{name}: unknown modern place')
        coordinates = island['coordinates']
        require(isinstance(coordinates, list) and len(coordinates) == 2 and
                all(type(n) in (int, float) and math.isfinite(n) for n in coordinates),
                f'{name}: coordinates must be two finite numbers [longitude, latitude]')
        lon, lat = coordinates
        require(-89 <= lon <= -58 and 9 <= lat <= 29, f'{name}: coordinates outside atlas bounds')
        control, sovereign = island['initialController'], island['initialSovereign']
        require(control in owner_ids and sovereign in owner_ids, f'{name}: unknown initial power')
        citations(island['sources'], name)
        require(isinstance(island['events'], list) and island['events'], f'{name}: missing events')
        previous_date, count = dt.date(meta['startYear'], 1, 1), 0
        for index, event in enumerate(island['events']):
            record(event, event_keys, f'{name}.events[{index}]', optional)
            eid = event['id']
            identifier(eid, f'{name}.events[{index}].id')
            require(eid.startswith(name + '-') and len(eid) > len(name) + 1 and eid not in all_event_ids,
                    f'Duplicate or invalid event ID: {eid}')
            all_event_ids.add(eid)
            try:
                when = date_key(event['date'])
            except ValueError as error:
                raise ValueError(f'{eid}.date: {error}') from error
            require(previous_date <= when <= dt.date(meta['endYear'], 12, 31),
                    f'{eid}: dates must be chronological and within atlas range')
            previous_date = when
            expected = {4: 'year', 7: 'month', 10: 'day'}[len(event['date'])]
            require(event['precision'] == expected or (event['precision'] == 'circa' and expected == 'year'),
                    f'{eid}: precision disagrees with date')
            require(isinstance(event['kind'], str) and event['kind'] in kinds, f'{eid}: unknown event kind')
            text_fields(event, ['title', 'detail'], eid)
            for field in optional:
                if field in event:
                    text(event[field], f'{eid}.{field}')
            require(event['precision'] != 'circa' or event.get('uncertainty'),
                    f'{eid}: circa date needs an uncertainty note')
            for field in ['controller', 'sovereign']:
                value = event[field]
                require(value is None or (isinstance(value, str) and value in owner_ids),
                        f'{eid}.{field}: unknown power')
            if event['kind'] in {'claim', 'context', 'status', 'resistance'}:
                require(event['controller'] is None and event['sovereign'] is None,
                        f'{eid}: an annotation cannot transfer control or title')
            if event['kind'] == 'claim' or 'claimant' in event:
                require(event.get('claimant') in owner_ids, f'{eid}: a claim needs a known claimant')
            citations(event['sources'], eid)
            event['year'] = when.year
            event['changesControl'] = event['controller'] is not None and event['controller'] != control
            event['changesSovereignty'] = event['sovereign'] is not None and event['sovereign'] != sovereign
            event['previousController'], event['previousSovereign'] = control, sovereign
            control, sovereign = event['controller'] or control, event['sovereign'] or sovereign
            event['resultingController'], event['resultingSovereign'] = control, sovereign
            count += event['changesControl']
        island['controlChanges'] = count
        island['currentController'], island['currentSovereign'] = control, sovereign
        islands.append(island)
    require({island['place'] for island in islands} == set(scope['places']), 'A modern place has no track')
    return dict(meta=meta, owners=owners, sources=sources, contexts=contexts, islands=islands)


def write_outputs(data):
    OUT.mkdir(parents=True, exist_ok=True)
    compact = json.dumps(data, ensure_ascii=False, separators=(',', ':')) + '\n'
    (OUT / 'caribbean.json').write_text(compact, encoding='utf-8')
    (ROOT / 'lib/caribbean.json').write_text(compact, encoding='utf-8')
    source_map = {s['id']: s for s in data['sources']}
    with (OUT / 'events.csv').open('w', newline='', encoding='utf-8') as file:
        keys = ['island', 'place', 'event_id', 'date', 'precision', 'kind', 'title', 'detail', 'controller', 'sovereign', 'changes_control', 'changes_sovereignty', 'uncertainty', 'qualification', 'source_ids', 'source_urls']
        writer = csv.DictWriter(file, fieldnames=keys)
        writer.writeheader()
        for island in data['islands']:
            for e in island['events']:
                writer.writerow(dict(island=island['name'], place=island['place'], event_id=e['id'], date=e['date'], precision=e['precision'], kind=e['kind'], title=e['title'], detail=e['detail'], controller=e['resultingController'], sovereign=e['resultingSovereign'], changes_control=e['changesControl'], changes_sovereignty=e['changesSovereignty'], uncertainty=e.get('uncertainty', ''), qualification=e.get('qualification', ''), source_ids='; '.join(e['sources']), source_urls='; '.join(source_map[s]['url'] for s in e['sources'])))
    notes = [dict(island=i['name'], event=e['id'], date=e['date'], issue=e['uncertainty']) for i in data['islands'] for e in i['events'] if e.get('uncertainty')]
    (OUT / 'editorial-notes.json').write_text(json.dumps(notes, ensure_ascii=False, indent=2) + '\n', encoding='utf-8')
    bibliography = ['# Bibliography', '', f"Compiled {data['meta']['compiled']}. See methodology.md for scope and editorial cautions.", '']
    for s in data['sources']:
        bibliography.append(f"- **{s['id']}** — {s['publisher']}. [{s['title']}]({s['url']}). {s['type']}; consulted {s['accessed']}.")
    (OUT / 'bibliography.md').write_text('\n'.join(bibliography) + '\n', encoding='utf-8')
    (OUT / 'methodology.md').write_bytes((DATA / 'methodology.md').read_bytes())


if __name__ == '__main__':
    try:
        compiled = compile_data()
        if '--check' not in sys.argv:
            write_outputs(compiled)
        print(f"Validated {len(compiled['islands'])} tracks, {sum(len(i['events']) for i in compiled['islands'])} events and {len(compiled['sources'])} sources.")
    except (ValueError, KeyError, TypeError) as error:
        sys.exit(f'Dataset validation failed: {error}')

File: scripts/build-data.mjs
import { mkdir, readFile, readdir, rm, writeFile } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const dataRoot = join(root, 'data');
const outputRoot = join(root, 'public/data');

function requireValue(condition, message) {
  if (!condition) throw new Error(message);
}

async function readJson(name) {
  try {
    const input = await readFile(join(dataRoot, name), 'utf8');
    const value = JSON.parse(input);
    // JSON.parse checks syntax, but silently overwrites duplicate object keys.
    // Consume whole strings so braces and colons inside prose are ignored.
    const objects = [];
    for (const [token] of input.matchAll(/"(?:\\.|[^"\\])*"\s*:?|[{}]/g)) {
      if (token === '{') objects.push(new Set());
      else if (token === '}') objects.pop();
      else if (token.endsWith(':')) {
        const key = JSON.parse(token.slice(0, -1));
        const keys = objects.at(-1);
        if (keys.has(key))
          throw new SyntaxError(`Duplicate key ${JSON.stringify(key)}`);
        keys.add(key);
      }
    }
    return value;
  } catch (error) {
    if (error instanceof SyntaxError)
      throw new Error(`${name}: ${error.message}`);
    throw error;
  }
}

function dateKey(value) {
  requireValue(
    typeof value === 'string' && /^\d{4}(-\d{2}){0,2}$/.test(value),
    `Invalid date: ${value}`,
  );
  const [year, month = 1, day = 1] = value.split('-').map(Number);
  const date = new Date(Date.UTC(year, month - 1, day));
  requireValue(
    date.getUTCFullYear() === year &&
      date.getUTCMonth() === month - 1 &&
      date.getUTCDate() === day,
    `Invalid date: ${value}`,
  );
  return date;
}

function record(value, required, context, optional = []) {
  requireValue(
    value !== null && typeof value === 'object' && !Array.isArray(value),
    `${context}: expected an object`,
  );
  const missing = required.filter((key) => !(key in value));
  const allowed = new Set([...required, ...optional]);
  const unknown = Object.keys(value).filter((key) => !allowed.has(key));
  requireValue(!missing.length, `${context}: missing fields ${missing.sort()}`);
  requireValue(
    !unknown.length,
    `${context}: unknown or generated fields ${unknown.sort()}`,
  );
}

function text(value, context) {
  requireValue(
    typeof value === 'string' && value.trim(),
    `${context}: expected nonempty text`,
  );
}

function textFields(value, names, context) {
  for (const name of names) text(value[name], `${context}.${name}`);
}

function identifier(value, context) {
  requireValue(
    typeof value === 'string' && /^[a-z][a-z0-9-]*$/.test(value),
    `${context}: expected a lowercase ID with letters, digits or hyphens`,
  );
}

function strings(value, context) {
  requireValue(
    Array.isArray(value) && value.length,
    `${context}: expected a nonempty list`,
  );
  for (const item of value) text(item, context);
  requireValue(
    new Set(value).size === value.length,
    `${context}: duplicate entry`,
  );
}

function year(value, context) {
  requireValue(
    Number.isInteger(value) && value >= 1000 && value <= 9998,
    `${context}: invalid year`,
  );
}

function records(value, required, context) {
  requireValue(
    Array.isArray(value) && value.length,
    `${context}: expected a nonempty list`,
  );
  const seen = new Set();
  value.forEach((item, index) => {
    const label = `${context}[${index}]`;
    record(item, required, label);
    identifier(item.id, `${label}.id`);
    requireValue(!seen.has(item.id), `${context}: duplicate ID ${item.id}`);
    seen.add(item.id);
  });
}

function csv(value) {
  const textValue = value == null ? '' : String(value);
  return /[",\n\r]/.test(textValue)
    ? `"${textValue.replaceAll('"', '""')}"`
    : textValue;
}

async function compileData() {
  const [meta, owners, sources, contexts, scope] = await Promise.all(
    ['meta', 'owners', 'sources', 'contexts', 'scope'].map((name) =>
      readJson(`${name}.json`),
    ),
  );
  const metaText = [
    'title',
    'version',
    'scope',
    'baseline',
    'method',
    'sovereignty',
    'completeness',
    'sourcePolicy',
    'license',
    'compiled',
    'currentStatusAsOf',
  ];
  record(meta, [...metaText, 'startYear', 'endYear'], 'meta.json');
  textFields(meta, metaText, 'meta.json');
  for (const field of ['compiled', 'currentStatusAsOf']) {
    requireValue(meta[field].length === 10, `meta.${field}: use YYYY-MM-DD`);
    dateKey(meta[field]);
  }
  for (const field of ['startYear', 'endYear'])
    year(meta[field], `meta.${field}`);
  requireValue(
    meta.startYear < meta.endYear,
    'meta: startYear must precede endYear',
  );

  records(owners, ['id', 'label', 'description', 'color'], 'owners.json');
  records(
    sources,
    ['id', 'title', 'publisher', 'type', 'url', 'accessed'],
    'sources.json',
  );
  records(
    contexts,
    ['id', 'start', 'end', 'title', 'description', 'sources'],
    'contexts.json',
  );
  const ownerIds = new Set(owners.map((owner) => owner.id));
  const sourceMap = new Map(sources.map((source) => [source.id, source]));
  for (const source of sources) {
    textFields(
      source,
      ['title', 'publisher', 'type', 'url', 'accessed'],
      source.id,
    );
    const url = new URL(source.url);
    requireValue(
      url.protocol === 'https:' && !url.username && !url.password,
      `${source.id}: source needs an HTTPS URL without credentials`,
    );
    requireValue(
      source.accessed.length === 10,
      `${source.id}: accessed needs YYYY-MM-DD`,
    );
    dateKey(source.accessed);
  }
  for (const owner of owners) {
    textFields(owner, ['label', 'description', 'color'], owner.id);
    requireValue(
      /^#[0-9a-fA-F]{6}$/.test(owner.color),
      `${owner.id}: invalid power colour`,
    );
  }
  const citations = (refs, context) => {
    strings(refs, `${context}.sources`);
    requireValue(
      refs.every((ref) => sourceMap.has(ref)),
      `${context}: unknown source in ${refs}`,
    );
  };
  for (const context of contexts) {
    const label = context.id;
    requireValue(
      !['all', 'custom'].includes(label),
      `${label}: reserved date-preset ID`,
    );
    textFields(context, ['title', 'description'], label);
    citations(context.sources, label);
    year(context.start, `${label}.start`);
    year(context.end, `${label}.end`);
    requireValue(
      meta.startYear <= context.start &&
        context.start < context.end &&
        context.end <= meta.endYear,
      `${label}: era must be within the atlas range`,
    );
  }
  record(scope, ['places', 'tracks'], 'scope.json');
  strings(scope.tracks, 'scope.tracks');
  strings(scope.places, 'scope.places');
  scope.tracks.forEach((name) => identifier(name, 'scope.tracks'));
  const files = new Set(
    (await readdir(join(dataRoot, 'islands')))
      .filter((name) => name.endsWith('.json'))
      .map((name) => name.slice(0, -5)),
  );
  requireValue(
    files.size === scope.tracks.length &&
      scope.tracks.every((track) => files.has(track)),
    'Island files and scope.json tracks do not match',
  );

  const islands = [];
  const allEventIds = new Set();
  const islandKeys = [
    'id',
    'name',
    'place',
    'region',
    'coordinates',
    'peoples',
    'summary',
    'notes',
    'sources',
    'initialController',
    'initialSovereign',
    'events',
  ];
  const eventKeys = [
    'id',
    'date',
    'precision',
    'kind',
    'title',
    'detail',
    'controller',
    'sovereign',
    'sources',
  ];
  const optional = ['uncertainty', 'qualification', 'claimant'];
  const kinds = new Set([
    'claim',
    'context',
    'resistance',
    'status',
    'capture',
    'restoration',
    'settlement',
    'treaty',
    'independence',
    'withdrawal',
    'context-change',
    'resistance-change',
  ]);
  for (const name of scope.tracks) {
    const island = await readJson(`islands/${name}.json`);
    record(island, islandKeys, name);
    requireValue(island.id === name, `${name}: filename and ID differ`);
    textFields(
      island,
      [
        'name',
        'place',
        'region',
        'peoples',
        'summary',
        'notes',
        'initialController',
        'initialSovereign',
      ],
      name,
    );
    requireValue(
      scope.places.includes(island.place),
      `${name}: unknown modern place`,
    );
    const coordinates = island.coordinates;
    requireValue(
      Array.isArray(coordinates) &&
        coordinates.length === 2 &&
        coordinates.every(
          (number) => typeof number === 'number' && Number.isFinite(number),
        ),
      `${name}: coordinates must be two finite numbers [longitude, latitude]`,
    );
    requireValue(
      coordinates[0] >= -89 &&
        coordinates[0] <= -58 &&
        coordinates[1] >= 9 &&
        coordinates[1] <= 29,
      `${name}: coordinates outside atlas bounds`,
    );
    let control = island.initialController;
    let sovereign = island.initialSovereign;
    requireValue(
      ownerIds.has(control) && ownerIds.has(sovereign),
      `${name}: unknown initial power`,
    );
    citations(island.sources, name);
    requireValue(
      Array.isArray(island.events) && island.events.length,
      `${name}: missing events`,
    );
    let previousDate = new Date(Date.UTC(meta.startYear, 0, 1));
    let count = 0;
    island.events.forEach((event, index) => {
      const context = `${name}.events[${index}]`;
      record(event, eventKeys, context, optional);
      identifier(event.id, `${context}.id`);
      requireValue(
        event.id.startsWith(`${name}-`) &&
          event.id.length > name.length + 1 &&
          !allEventIds.has(event.id),
        `Duplicate or invalid event ID: ${event.id}`,
      );
      allEventIds.add(event.id);
      let when;
      try {
        when = dateKey(event.date);
      } catch (error) {
        throw new Error(`${event.id}.date: ${error.message}`);
      }
      requireValue(
        previousDate <= when &&
          when <= new Date(Date.UTC(meta.endYear, 11, 31)),
        `${event.id}: dates must be chronological and within atlas range`,
      );
      previousDate = when;
      const expected = { 4: 'year', 7: 'month', 10: 'day' }[event.date.length];
      requireValue(
        event.precision === expected ||
          (event.precision === 'circa' && expected === 'year'),
        `${event.id}: precision disagrees with date`,
      );
      requireValue(
        typeof event.kind === 'string' && kinds.has(event.kind),
        `${event.id}: unknown event kind`,
      );
      textFields(event, ['title', 'detail'], event.id);
      for (const field of optional)
        if (field in event) text(event[field], `${event.id}.${field}`);
      requireValue(
        event.precision !== 'circa' || event.uncertainty,
        `${event.id}: circa date needs an uncertainty note`,
      );
      for (const field of ['controller', 'sovereign'])
        requireValue(
          event[field] === null ||
            (typeof event[field] === 'string' && ownerIds.has(event[field])),
          `${event.id}.${field}: unknown power`,
        );
      if (['claim', 'context', 'status', 'resistance'].includes(event.kind))
        requireValue(
          event.controller === null && event.sovereign === null,
          `${event.id}: an annotation cannot transfer control or title`,
        );
      if (event.kind === 'claim' || 'claimant' in event)
        requireValue(
          ownerIds.has(event.claimant),
          `${event.id}: a claim needs a known claimant`,
        );
      citations(event.sources, event.id);
      event.year = when.getUTCFullYear();
      event.changesControl =
        event.controller !== null && event.controller !== control;
      event.changesSovereignty =
        event.sovereign !== null && event.sovereign !== sovereign;
      event.previousController = control;
      event.previousSovereign = sovereign;
      control = event.controller || control;
      sovereign = event.sovereign || sovereign;
      event.resultingController = control;
      event.resultingSovereign = sovereign;
      count += Number(event.changesControl);
    });
    island.controlChanges = count;
    island.currentController = control;
    island.currentSovereign = sovereign;
    islands.push(island);
  }
  requireValue(
    new Set(islands.map((island) => island.place)).size ===
      scope.places.length &&
      scope.places.every((place) =>
        islands.some((island) => island.place === place),
      ),
    'A modern place has no track',
  );
  return { meta, owners, sources, contexts, islands };
}

async function writeOutputs(compiled) {
  await rm(outputRoot, { recursive: true, force: true });
  await mkdir(outputRoot, { recursive: true });
  const compact = `${JSON.stringify(compiled)}\n`;
  await writeFile(join(outputRoot, 'caribbean.json'), compact);
  await writeFile(join(root, 'src/lib/caribbean.json'), compact);
  const sourceMap = new Map(
    compiled.sources.map((source) => [source.id, source]),
  );
  const keys = [
    'island',
    'place',
    'event_id',
    'date',
    'precision',
    'kind',
    'title',
    'detail',
    'controller',
    'sovereign',
    'changes_control',
    'changes_sovereignty',
    'uncertainty',
    'qualification',
    'source_ids',
    'source_urls',
  ];
  const eventRows = compiled.islands.flatMap((island) =>
    island.events.map((event) => [
      island.name,
      island.place,
      event.id,
      event.date,
      event.precision,
      event.kind,
      event.title,
      event.detail,
      event.resultingController,
      event.resultingSovereign,
      event.changesControl,
      event.changesSovereignty,
      event.uncertainty || '',
      event.qualification || '',
      event.sources.join('; '),
      event.sources.map((source) => sourceMap.get(source).url).join('; '),
    ]),
  );
  const rows = [keys, ...eventRows.map((row) => row.map(csv).join(','))];
  await writeFile(join(outputRoot, 'events.csv'), `${rows.join('\n')}\n`);
  const notes = compiled.islands.flatMap((island) =>
    island.events
      .filter((event) => event.uncertainty)
      .map((event) => ({
        island: island.name,
        event: event.id,
        date: event.date,
        issue: event.uncertainty,
      })),
  );
  await writeFile(
    join(outputRoot, 'editorial-notes.json'),
    `${JSON.stringify(notes, null, 2)}\n`,
  );
}

try {
  const compiled = await compileData();
  if (!process.argv.includes('--check')) await writeOutputs(compiled);
  console.log(
    `Validated ${compiled.islands.length} tracks, ${compiled.islands.reduce((count, island) => count + island.events.length, 0)} events and ${compiled.sources.length} sources.`,
  );
} catch (error) {
  console.error(`Dataset validation failed: ${error.message}`);
  process.exitCode = 1;
}
